relay-quotas/src/scoping.rs

```diff
@@ -0,0 +1,343 @@
+/// copyable type intended for the use with `RateLimits` and `RateLimiter`. It implements
+/// `Deref<Target = Scoping>` and `AsRef<Scoping>` for ease of use.
+#[derive(Clone, Copy, Debug)]
+pub struct ItemScoping<'a> {
+    /// The data category of the item.
+    pub category: DataCategory,
+
+    /// Scoping of the data.
+    pub scoping: &'a Scoping,
+}
+
+impl AsRef<Scoping> for ItemScoping<'_> {
+    fn as_ref(&self) -> &Scoping {
+        self.scoping
+    }
+}
+
+impl std::ops::Deref for ItemScoping<'_> {
+    type Target = Scoping;
+
+    fn deref(&self) -> &Self::Target {
+        self.scoping
+    }
+}
+
+impl ItemScoping<'_> {
+    /// Returns the identifier of the given scope.
+    pub fn scope_id(&self, scope: QuotaScope) -> Option<u64> {
+        match scope {
+            QuotaScope::Organization => Some(self.organization_id),
+            QuotaScope::Project => Some(self.project_id.value()),
+            QuotaScope::Key => self.key_id,
+            QuotaScope::Unknown => None,
+        }
+    }
+
+    /// Checks whether the category matches any of the quota's categories.
+    pub fn matches_categories(&self, categories: &DataCategories) -> bool {
+        // An empty list of categories means that this quota matches all categories. Note that we
+        // skip `Unknown` categories silently. If the list of categories only contains `Unknown`s,
+        // we do **not** match, since apparently the quota is meant for some data this Relay does
+        // not support yet.
+        categories.is_empty() || categories.iter().any(|cat| *cat == self.category)
+    }
+
+    /// Checks whether this item scoping matches the given quota's scope.
+    ///
+    /// This quota matches, if:
+    ///  - there is no `scope_id` constraint
+    ///  - the `scope_id` constraint is not numeric
+    ///  - the scope identifier matches the one from ascoping and the scope is known
+    fn matches_quota_scope(&self, quota: &Quota) -> bool {
+        // Check for a scope identifier constraint. If there is no constraint, this means that the
+        // quota matches any scope. In case the scope is unknown, it will be coerced to the most
+        // specific scope later.
+        let scope_id = match quota.scope_id {
+            Some(ref scope_id) => scope_id,
+            None => return true,
+        };
+
+        // Check if the scope identifier in the quota is parseable. If not, this means we cannot
+        // fulfill the constraint, so the quota does not match.
+        let parsed = match scope_id.parse::<u64>() {
+            Ok(parsed) => parsed,
+            Err(_) => return false,
+        };
+
+        // At this stage, require that the scope is known since we have to fulfill the constraint.
+        self.scope_id(quota.scope) == Some(parsed)
+    }
+
+    /// Checks whether the quota's constraints match the current item.
+    pub fn matches(&self, quota: &Quota) -> bool {
```

Comment:
This used to be a method on `Quota`, now it's a method on `ItemScoping`. So instead of 

```rust
quota.matches(&item_scoping)
````

we now write

```rust
item_scoping.matches(&quota)
```

This way, `ItemScoping` can remain part of `relay-quotas`.